x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java

```diff
@@ -649,142 +711,270 @@ private void indexInvalidation(Collection<String> tokenIds, ActionListener<Token
+        final Optional<ElasticsearchSecurityException> invalidSource = checkTokenDocForRefresh(source, clientAuth);
+        if (invalidSource.isPresent()) {
+            onFailure.accept(invalidSource.get());
         } else {
-            Consumer<Exception> onFailure = ex -> listener.onFailure(traceLog("refresh token", tokenDocId, ex));
-            GetRequest getRequest = client.prepareGet(SecurityIndexManager.SECURITY_INDEX_NAME, TYPE, tokenDocId).request();
-            executeAsyncWithOrigin(client.threadPool().getThreadContext(), SECURITY_ORIGIN, getRequest,
-                ActionListener.<GetResponse>wrap(response -> {
-                    if (response.isExists()) {
-                        final Map<String, Object> source = response.getSource();
-                        final Optional<ElasticsearchSecurityException> invalidSource = checkTokenDocForRefresh(source, clientAuth);
-
-                        if (invalidSource.isPresent()) {
-                            onFailure.accept(invalidSource.get());
+            if (eligibleForMultiRefresh(source, refreshRequested)) {
+                final Map<String, Object> refreshTokenSrc = (Map<String, Object>) source.get("refresh_token");
+                final String supersedingTokenDocId = (String) refreshTokenSrc.get("superseded_by");
+                logger.debug("Token document [{}] was recently refreshed, attempting to reuse [{}] for returning an " +
+                    "access token and refresh token", tokenDocId, supersedingTokenDocId);
+                final ActionListener<GetResponse> getSupersedingListener = new ActionListener<GetResponse>() {
+                    @Override
+                    public void onResponse(GetResponse response) {
+                        if (response.isExists()) {
+                            logger.debug("Found superseding token document [{}] ", supersedingTokenDocId);
+                            final Map<String, Object> supersedingTokenSource = response.getSource();
+                            final Map<String, Object> supersedingUserTokenSource = (Map<String, Object>)
+                                ((Map<String, Object>) supersedingTokenSource.get("access_token")).get("user_token");
+                            final Map<String, Object> supersedingRefreshTokenSrc =
+                                (Map<String, Object>) supersedingTokenSource.get("refresh_token");
+                            final String supersedingRefreshTokenValue = (String) supersedingRefreshTokenSrc.get("token");
+                            reIssueTokens(supersedingUserTokenSource, supersedingRefreshTokenValue, listener);
+                        } else if (backoff.hasNext()) {
+                            // We retry this since the creation of the superseding token document might already be in flight but not
+                            // yet completed, triggered by a refresh request that came a few milliseconds ago
```

Comment:
I would've created the new refreshed user token first and then updated the pointer.
This way we avoid the back-off. If we get a concurrent modification exception then we can retry and _assert_ that we have a valid refresh token concurrently generated (any extraneous user tokens can be removed or garbage collected).

**No need to change the code for this, can be done as a follow-up.**

Reply:
The idea behind doing the update first is that I focused on being able to handle the subsequent requests (after a few millis ) so it made sense to do the update so that next ones can see that immediately. If I get this right, your idea would mean that roughly:  

- X : 1st request received. Start the process to index the superseding doc
- X+y ms :  2nd request received. The previous index request is in flight, so we start the process to index the (new and different) superseding doc
- X+z ms : 1st index request completes, we update the original document with the pointer to the superseding doc and reply to the API request with the newly created refresh token
- X+ω ms : 2nd index request completes, we try to update the original doc but we get a version conflict exception, so we do a Get request now to get the pointer to the superseding doc and then Get that one to get it's source and reply with the same refresh token

The disadvantage of that are the extra refresh and access token we create. In addition to space ( which is a short term concern since as you said we clean up the tokens) what worries me a little is that we raise the possibilities of a successful brute force attack by introducing unnecessarily more valid tokens. The time validity of the access tokens and the key space for our access tokens and refresh tokens, makes this effect insignificant but if we don't get much in terms of performance/complexity minimization then I'm not sure if it's worth following. I'm definitely up to continue the discussion on it though!

Reply:
Yes, you're right, this trades a bit of space for complexity.
I think we can alleviate this by having the id of the refreshed user token be a hash of the user token they are refreshing (superseding) - a hash of the id and creation time. This way concurrent refreshes would be trying to create the same doc_id and fail, only one will succeed and the others will retry, picking up the only refreshed token.

Reply:
> Yes, you're right, this trades a bit of space for complexity.

And , _slightly_, robustness against brute force attacks. 

>  I think we can alleviate this by having the id of the refreshed user token be a hash of the user token they are refreshing (superseding) - a hash of the id and creation time.

I don't know how I feel about making the generation of token document IDs ( == access tokens ) predictable. I'll think about this a bit more